src/bin/ion/commands/generate/generator.rs

```diff
@@ -212,118 +333,246 @@ impl<'a, L: Language + 'static> CodeGenerator<'a, L> {
-        let mut tera_fields = vec![];
+        field_presence: FieldPresence,
+        parent_code_gen_context: &mut CodeGenContext,
+    ) -> CodeGenResult<FullyQualifiedTypeReference> {
         let mut code_gen_context = CodeGenContext::new();
-        let mut nested_anonymous_types = vec![];
-        let constraints = isl_type.constraints();
-        for constraint in constraints {
-            self.map_constraint_to_abstract_data_type(
-                &mut nested_anonymous_types,
-                &mut tera_fields,
-                constraint,
-                &mut code_gen_context,
-            )?;
-        }
+        let mut data_model_node = self.convert_isl_type_def_to_data_model_node(
+            type_name,
+            isl_type,
+            &mut code_gen_context,
+            true,
+        )?;
 
-        // TODO: verify the `occurs` value within a field, by default the fields are optional.
-        if let Some(abstract_data_type) = &code_gen_context.abstract_data_type {
-            // Add the nested type into parent type's tera context
-            nested_types.push(NestedType {
-                target_kind_name: type_name.to_case(Case::UpperCamel),
-                fields: tera_fields,
-                abstract_data_type: abstract_data_type.to_owned(),
-                nested_types: nested_anonymous_types,
-            });
-        } else {
-            return invalid_abstract_data_type_error(
-                "Can not determine abstract data type, specified constraints do not map to an abstract data type.",
-            );
+        // add this nested type to parent code gene context's current list of nested types
+        parent_code_gen_context
+            .nested_types
+            .push(data_model_node.to_owned());
+
+        // since nested sequence does not create a separate class, all its nested types should also be added to parent code gen context
+        if data_model_node.is_sequence() {
+            parent_code_gen_context
+                .nested_types
+                .extend_from_slice(&data_model_node.nested_types);
         }
 
-        Ok(())
+        match field_presence {
+            FieldPresence::Optional => Ok(L::target_type_as_optional(
+                data_model_node.fully_qualified_type_ref::<L>().ok_or(
+                    invalid_abstract_data_type_raw_error(
+                        "Can not determine fully qualified name for the data model",
+                    ),
+                )?,
+            )),
+            FieldPresence::Required => data_model_node.fully_qualified_type_ref::<L>().ok_or(
+                invalid_abstract_data_type_raw_error(
+                    "Can not determine fully qualified name for the data model",
+                ),
+            ),
+        }
     }
 
     fn generate_abstract_data_type(
         &mut self,
-        isl_type_name: &str,
+        isl_type_name: &String,
```

Comment:
(🗺️ PR tour) This was changed due to merge where newly introduced method `convert_isl_type_def_to_data_model_node` started using `&String`